packages/daimo-api/src/contract/SwapClogMatcher.ts

```diff
@@ -0,0 +1,115 @@
+import { Pool } from "pg";
+import { Address, bytesToHex, getAddress } from "viem";
+
+import { ForeignTokenTransfer } from "./foreignCoinIndexer";
+import { TokenRegistry } from "../server/tokenRegistry";
+import { retryBackoff } from "../utils/retryBackoff";
+
+export class SwapClogMatcher {
+  /** Map of transactionHash to transfers that are within that transaction */
+  private txHashToTransfers: Map<string, ForeignTokenTransfer[]> = new Map();
+
+  /** Map of starting "to" transfer to the corresponding "from" transfer */
+  private correspondingSwapTransfers: Map<
+    ForeignTokenTransfer,
+    ForeignTokenTransfer
+  > = new Map();
+
+  constructor(private tokenReg: TokenRegistry) {}
+
+  // Given a set transaction hashes, load any transfers that are associated
+  // with those transactions.
+  async loadSwapTransfers(
+    pg: Pool,
+    from: number,
+    to: number,
+    chainId: number,
+    transactionHashes: string[]
+  ) {
+    const startTime = Date.now();
+
+    const txHashes = transactionHashes.map((tx) => `\\x${tx.substring(2)}`);
+
+    const result = await retryBackoff(
+      `swapClogMatcher-logs-query-${from}-${to}`,
+      async () => {
+        return await pg.query(
+          `SELECT * from erc20_transfers 
+          WHERE (block_num BETWEEN $1 AND $2) 
+          AND chain_id = $3
+          AND tx_hash = ANY($4::bytea[]);`,
+          [from, to, chainId, txHashes as any]
```

Comment:
create index on erc20_transfers for (block_num, chain_id, tx_hash)

Reply:
add to `migrateDB()` in watcher

Reply:
alternatively, since tx_hash uniquely specifies both block number, it may work best to index by just tx_hash:

```sql
CREATE INDEX IF NOT EXISTS i_erc20_transfers_tx_hash ON erc20_transfers (tx_hash);
```